[
  {
    "number": 30128,
    "title": "Partial Import does not work",
    "created_at": "2024-06-04T09:09:43Z",
    "closed_at": "2024-06-06T07:24:45Z",
    "labels": [
      "kind/bug",
      "area/admin/ui",
      "help wanted",
      "team/ui",
      "priority/low",
      "status/auto-expire",
      "status/auto-bump"
    ],
    "url": "https://github.com/keycloak/keycloak/issues/30128",
    "body": "### Before reporting an issue\n\n- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.\n\n### Area\n\nadmin/ui\n\n### Describe the bug\n\nAfter using \"Partial Export\" I cannot import the file using \"Partial Import\" (the \"Import\" button stays grayed out).\n\n### Version\n\n24.0.5\n\n### Regression\n\n- [ ] The issue is a regression\n\n### Expected behavior\n\nfile can be imported (by browse, d'n'd, paste)\n\n### Actual behavior\n\n\"Import\" button stays grayed out.\n\n### How to Reproduce?\n\n1. create new realm\r\n2. go to realm settings\r\n3. actions --> partial export: keep defaults, save file.\r\n4. actions --> partial import: browser for file\r\n\r\nobserve the import button stays grayed out.\n\n### Anything else?\n\ntested in Firefox 126 and Edge 125\r\n\r\ntested via container: `podman run -p 8080:8080 -e KEYCLOAK_ADMIN=admin -e KEYCLOAK_ADMIN_PASSWORD=admin quay.io/keycloak/keycloak:24.0.5 start-dev`, both 24.0.5 (shown) and nightly (not shown, I made sure to pull latest nightly).",
    "comments_url": "https://api.github.com/repos/keycloak/keycloak/issues/30128/comments",
    "author": "ciis0",
    "comments": [
      {
        "user": "ssilvert",
        "created_at": "2024-06-05T19:21:58Z",
        "body": "The reason that the `Import` button is not active is because there is nothing to import.\r\n\r\nA partial import can import users, clients, identity providers, groups, and roles.  None of those exist in the partial export when you do an export on a new realm with all defaults.\r\n\r\nWe probably should have a message for this to make it more clear though."
      },
      {
        "user": "ciis0",
        "created_at": "2024-06-06T07:24:45Z",
        "body": "ah, ok, i was trying to import client scopes... :)"
      },
      {
        "user": "ciis0",
        "created_at": "2024-06-06T07:43:04Z",
        "body": "(I accidentally deleted the OpenID `profile` scope and wanted to restore it via that... ^^)"
      },
      {
        "user": "jaeaxt",
        "created_at": "2024-12-02T13:00:04Z",
        "body": "Hi, I'm trying to do a partial import for organization but it doesn't work. The import button is not active even when I copy and past the exported data. But If use the option import-realm at start with adding the organizations to my realm file it's work."
      }
    ]
  },
  {
    "number": 27984,
    "title": "Username LDAP attribute other than uid is difficult",
    "created_at": "2024-03-18T10:45:24Z",
    "closed_at": "2024-04-03T12:11:56Z",
    "labels": [
      "kind/bug",
      "area/ldap",
      "help wanted",
      "team/core-iam",
      "status/auto-expire",
      "status/auto-bump",
      "priority/normal",
      "release/25.0.0"
    ],
    "url": "https://github.com/keycloak/keycloak/issues/27984",
    "body": "### Before reporting an issue\r\n\r\n- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.\r\n\r\n### Area\r\n\r\nldap\r\n\r\n### Describe the bug\r\n\r\nI am trying to  create a new User Federation LDAP provider, using firstname.lastname as the user login scheme. So, on my OpenLDAP server, I tried several usually empty LDAP fields where to enter firtsname.lastname for my test user, for example title, employeeNumber, and postOfficeBox. But none of these works as Username LDAP attribute in Keycloak and syncing users fails. The log shows:\r\n\r\nERROR [org.keycloak.storage.ldap.LDAPStorageProviderFactory] (executor-thread-95) Failed during import user from LDAP: org.keycloak.models.ModelException: User returned from LDAP has null username! Check configuration of your LDAP mappings.\r\n\r\n### Version\r\n\r\nlatest\r\n\r\n### Regression\r\n\r\n- [ ] The issue is a regression\r\n\r\n### Expected behavior\r\n\r\nKeycloak syncs users according to the newly set Username LDAP attribute\r\n\r\n### Actual behavior\r\n\r\nAn error message is shown which says that no user could be imported\r\n\r\n### How to Reproduce?\r\n\r\nOn the OpenLDAP server, set firstname.lastname as employeeNumber, title and postOfficeBox for some test user. Set one of these as User LDAP attribute in Keycloak and then sync all users.\r\n\r\n### Anything else?\r\n\r\n_No response_",
    "comments_url": "https://api.github.com/repos/keycloak/keycloak/issues/27984/comments",
    "author": "Cistoge",
    "comments": [
      {
        "user": "sguilhen",
        "created_at": "2024-03-19T21:48:17Z",
        "body": "I've tried to reproduce the problem with OpenLDAP, and setting `cn` to the username attribute worked fine - users were updated and synch succeeded. I've tried even to use `uid` first, import the users, then switch it to `cn` and re-synch. All worked fine."
      },
      {
        "user": "Cistoge",
        "created_at": "2024-03-20T10:22:45Z",
        "body": "Hm, maybe I made mistake on monday and forgot to save my settings after changing from Username LDAP attribute uidNumber to cn, for now it works with cn also for me. But other fields like uidNumber or employeeNumber definitely do not work for me. "
      },
      {
        "user": "sguilhen",
        "created_at": "2024-03-27T12:53:45Z",
        "body": "In my setup, I was able to use other attributes as well, such as `employeeNumber`, and import worked fine. If one of the users doesn't have this attribute, it will skip the user during import - in the logs you can see the same message you've posted about the username being `null` (admin console says 1 user failed during synch). So whatever attribute is chosen as the username attribute, all entries in LDAP must have it set or Keycloak won't accept importing the user.\r\n\r\nOne thing I've noticed, though, is that while synch works and reports some users failed to be imported (those with `null` values retrieved from LDAP for the username), when searching for all users in the Users screen the whole thing fails if an user from LDAP has a `null` username, instead of skipping them. We might need to revisit the `searchForUserStream` in `LDAPStorageProvider` to keep the behavior consistent."
      },
      {
        "user": "Cistoge",
        "created_at": "2024-03-27T15:19:31Z",
        "body": "The OpenLDAP instance I am using for this is not for production use, but only for testing and currently has only one user.  This user has \"testuser\" as uid OpenLDAP attribute and \"test.user\" as postOfficeBox OpenLDAP attribute. In Keycloak, the user can be imported with uid as Username LDAP attribute, but not with postOfficeBox ..."
      },
      {
        "user": "sguilhen",
        "created_at": "2024-03-28T17:50:27Z",
        "body": "I think I see what's going on. When I create a LDAP federation from scratch, I can set anything to the Username LDAP attribute, and synch users works as expected.\r\n\r\nNow, if we go back to the already configured provider, and edit the Username LDAP attribute (for example, switch from `uid` to `postOfficeBox `), I start getting the same errors, even after removing all previously imported users. So here's what is happening:\r\n\r\n1. When we first create an LDAP storage provider, the configured Username LDAP attribute (let's say `uid`), is used to create a `username` mapper that fetches the same attribute from LDAP when loading users, and sets it as the username in Keycloak.\r\n2.  When we edit the Username LDAP attribute in the LDAP provider (let's say to `postOfficeBox`), the change **is not** propagated to the `username` mapper, which is still with the old `uid` value. Without any mappers configured to fetch the `postOfficeBox` attribute from LDAP, the value is read as `null` when importing, giving us the errors you've described.\r\n\r\nIf, after changing the Username LDAP attribute in the main LDAP config you also edit the `username` mapper to the same attribute, synchronizing users work again, and any previously imported user will be updated to the new username.\r\n\r\nIt would be good if the change was propagated to the correct mapper once the LDAP config changes, but for now this is a workaround you can use."
      }
    ]
  }
]